src/vsr.zig

```diff
@@ -1917,3 +1918,52 @@ const ViewChangeHeadersArray = struct {
         headers.array.appendAssumeCapacity(Headers.dvc_blank(op));
     }
 };
+
+pub const Op = struct {
+    pub fn checkpoint_before(checkpoint: u64) ?u64 {
+        assert(checkpoint_valid(checkpoint));
+
+        if (checkpoint == 0) {
+            return null;
+        }
+
+        if (checkpoint == constants.journal_slot_count - constants.lsm_batch_multiple - 1) {
```

Comment:
What are our local expectations here w.r.t. the relation between `journal_slot_count` and `lsm_batch_multiple`?

We should assert them here, regardless of where else they are asserted, as defense in depth.

Reply:
If you think it's necessary, we could comment to the effect that `checkpoint` is an index, whereas slot count and multiple are counts, therefore `- 1` to convert count to index.

Reply:
Since this distance is also shared by both `before` and `after` (we do the same subtraction in a few places), we could use a named local constant instead of the magic calculation.

Reply:
Then it could become `+ ops_between_checkpoints` or `- ops_between_checkpoints`.

Reply:
I'd prefer to leave that constant inlined, since "ops_between_checkpoints" is misleading -- it is really "ops between checkpoints except for between the origin and next checkpoint". (The transition from "checkpoint=0" to its next checkpoint is a special case.)

Reply:
It does seem that the subtraction is performed in enough places that it could be factored out, perhaps with a better name.